js/private/js_image_layer.bzl

```diff
@@ -254,210 +262,295 @@ container_image(
-        entry = {
-            "dest": file.path,
-            "root": file.root.path,
-            "is_external": file.owner.repo_name != "",
-            "is_source": file.is_source,
-            "is_directory": file.is_directory,
-        }
-        if destination == real_binary_path:
-            entry["remove_non_hermetic_lines"] = True
-
-        all_entries[destination] = entry
-
-        layer = _select_layer(layers, destination, file)
-        layer.entries[destination] = entry
-        layer.inputs.append(file)
+    # shell launcher generated by js_binary contains non-reproducible information swap it out with the sanitized one.
+    entries.add_joined(
+        [runfiles_dir + "/" + _to_rlocation_path(binary_default_info.files_to_run.executable, workspace_name), {"dest": launcher.path, "root": launcher.root.path}],
+        join_with = ":",
+        map_each = json.encode,
+    )
 
     if repo_mapping:
-        destination = paths.join(runfiles_dir, "_repo_mapping")
-        entry = {
-            "dest": repo_mapping.path,
-            "root": repo_mapping.root.path,
-        }
-        all_entries[destination] = entry
-        layers.app.entries[destination] = entry
-
-    all_entries_json = ctx.actions.declare_file("{}_all_entries.json".format(ctx.label.name))
-    ctx.actions.write(all_entries_json, content = json.encode(all_entries))
-
-    node = _build_layer(
-        ctx,
-        type = "node",
-        all_entries_json = all_entries_json,
-        entries = layers.node.entries,
-        inputs = layers.node.inputs,
-    )
-    package_store_3p = _build_layer(
-        ctx,
-        type = "package_store_3p",
-        all_entries_json = all_entries_json,
-        entries = layers.package_store_3p.entries,
-        inputs = layers.package_store_3p.inputs,
-    )
-    package_store_1p = _build_layer(
-        ctx,
-        type = "package_store_1p",
-        all_entries_json = all_entries_json,
-        entries = layers.package_store_1p.entries,
-        inputs = layers.package_store_1p.inputs,
-    )
-    node_modules = _build_layer(
-        ctx,
-        type = "node_modules",
-        all_entries_json = all_entries_json,
-        entries = layers.node_modules.entries,
-        inputs = layers.node_modules.inputs,
-    )
-    app = _build_layer(
-        ctx,
-        type = "app",
-        all_entries_json = all_entries_json,
-        entries = layers.app.entries,
-        inputs = layers.app.inputs,
-    )
+        entries.add(",")
+        entries.add_joined(
+            [runfiles_dir + "/" + "_repo_mapping", {"dest": repo_mapping.path, "root": repo_mapping.root.path}],
+            join_with = ":",
+            map_each = json.encode,
+        )
+    entries.add("}")
+
+    entries_json = ctx.actions.declare_file("{}_entries.json".format(ctx.label.name))
+    ctx.actions.write(entries_json, content = entries)
+
+    # Ordering of these matter.
+    layer_groups = dict()
+    for key in ctx.attr.layer_groups:
```

Comment:
@alexeagle do you have any opinions on this API? I haven't used `js_image_layer` enough to have a good opinion here. I find it a bit odd that custom layers go before the node layer, but I can't think of any better API to overcome that.